packages/moonbit-async.mbt.md

```diff
@@ -0,0 +1,1267 @@
+  // WARNING: This API is for testing only
+  let tcp_server = @socket.TCPServer::new(@socket.Addr::parse("0.0.0.0:8443"))
+  defer tcp_server.close()
+
+  @async.with_task_group(fn(group) {
+    for {
+      let (tcp_conn, addr) = tcp_server.accept()
+      println("TLS connection from: \{addr}")
+
+      group.spawn_bg(allow_failure=true, fn() {
+        defer tcp_conn.close()
+
+        let tls_conn = @tls.TLS::server(
+          tcp_conn,
+          private_key_file="server.key",
+          private_key_type=PEM,
+          certificate_file="server.crt",
+          certificate_type=PEM
+        )
+
+        // Handle TLS connection
+        // handle_tls_request(tls_conn)
+
+        @async.with_timeout_opt(5000, fn() {
+            tls_conn.shutdown()
+          }) |> ignore
+          tls_conn.close()
+      })
+    }
+  })
+}
+```
+
+Notice that async function is not supported in `defer` yet.
+
+## HTTP Programming
+
+### HTTP File Server
+
+```moonbit
+///|
+async fn serve_file(
+  conn : @http.ServerConnection,
+  file : @fs.File,
+  path~ : String,
+) -> Unit {
+  let content_type = match path {
+    [.., .. ".html"] => "text/html"
+    [.., .. ".css"] => "text/css"
+    [.., .. ".js"] => "text/javascript"
+    [.., .. ".png"] => "image/png"
+    [.., .. ".jpg"] => "image/jpeg"
+    _ => "application/octet-stream"
+  }
+
+  conn
+  ..send_response(200, "OK", extra_headers={ "Content-Type": content_type })
+  ..write_reader(file)
+  ..end_response()
+}
+
+///|
+async fn serve_directory(
+  conn : @http.ServerConnection,
+  dir : @fs.Directory,
+  path~ : String,
+) -> Unit {
+  let files = dir.read_all()
+  files.sort()
+  conn
+  ..send_response(200, "OK", extra_headers={ "Content-Type": "text/html" })
+  ..write("<!DOCTYPE html><html><head></head><body>")
+  ..write_string("<h1>\{path}</h1>\n", encoding=UTF8)
+  ..write("<div style=\"margin: 1em; font-size: 15pt\">")
+  ..write_string(
+    "<a href=\"\{path}?download_zip\">download as zip</a><br/><br/>\n",
+    encoding=UTF8,
+  )
+  if path[:-1].rev_find("/") is Some(index) {
+    let parent : StringView = if index == 0 { "/" } else { path[:index] }
+    conn.write_string("<a href=\"\{parent}\">..</a><br/><br/>\n", encoding=UTF8)
+  }
+  for file in files {
+    let sep = if path[path.length() - 1] != '/' { "/" } else { "" }
+    conn.write_string(
+      "<a href=\"\{path}\{sep}\{file}\">\{file}</a><br/>\n",
+      encoding=UTF8,
+    )
+  }
+  conn..write("</div></body></html>")..end_response()
+}
+
+///|
+async fn http_file_server() -> Unit {
+  let base_path = "."
+  let server = @socket.TCPServer::new(@socket.Addr::parse("0.0.0.0:8111"))
+  defer server.close()
+
+  println("HTTP file server started on port 8111")
+
+  @async.with_task_group(fn(ctx) {
+    for {
+      let (tcp_conn, addr) = server.accept()
+      let conn = @http.ServerConnection::new(tcp_conn)
+      println("New connection from: \{addr}")
+
+      ctx.spawn_bg(allow_failure=true, fn() {
```

Comment:
there's a utility `@http.run_server` that handle the loop structure automatically now